### PR TITLE
Deduper

### PR DESCRIPTION
**Pull request by smellslikeml**
Implements a Transformer for image deduplication/clustering.

Transforms DataFrame with column of image uris to add column of cluster ids based on hierarchical clustering of grouped images based on SSIM.

`groupIdCol` is used to reduce similarity computations to smaller sets of images likely to be similar. Pinterest used ids from [LSH hashed embeddings](https://medium.com/pinterest-engineering/detecting-image-similarity-using-spark-lsh-and-tensorflow-618636afc939) but the ids can come from hashed metadata such as movie IMDb id.

### CODE REVIEW

File: python/rikai/spark/functions/vision.py [hunk 1]
```diff
@@ -280,3 +281,22 @@ def spectrogram_image(
         format=image_format,
         **image_kwargs,
     )
+
+
+@udf(returnType=ArrayType(IntegerType()))
```
Comment (changhiskhan) [inline python/rikai/spark/functions/vision.py:286]: let's move this to rikai.contrib.spark.vision maybe? how do you feel about putting functionality that's useful but not quite so common to most rikai users to the contrib section @eddyxu ?

File: python/rikai/spark/functions/vision.py [hunk 2]
```diff
@@ -280,3 +281,22 @@ def spectrogram_image(
...
+def img_cluster(img_list, threshold=0.5, img_shape=(300, 300)):
```
Comment (changhiskhan) [inline python/rikai/spark/functions/vision.py:287]: @smellslikeml does keyword arguments work? like you can't do `img_cluster("image_col", img_shape=(400, 300))` can you?

Comment (changhiskhan) [inline python/rikai/spark/functions/vision.py:287]: some documentation would be nice, especially describing the return value

File: python/rikai/spark/functions/vision.py [hunk 3]
```diff
@@ -280,3 +281,22 @@ def spectrogram_image(
...
+    from sklearn.cluster import AgglomerativeClustering
+    from skimage.metrics import structural_similarity as ssim
+
+    if len(img_list) == 1:
+        return [0]
+    clustering_model = AgglomerativeClustering(
+        n_clusters=None, distance_threshold=threshold, linkage="ward"
+    )
+    frames = [np.resize(img.to_numpy(), img_shape) for img in img_list]
```
Comment (changhiskhan) [inline python/rikai/spark/functions/vision.py:296]: from documentation on numpy.resize: "This functionality is therefore not suitable to resize images, or data where each axis represents a separate and distinct entity."

Comment (changhiskhan) [reply python/rikai/spark/functions/vision.py:296]: why call this "frames" as opposed to just images?

File: python/rikai/spark/functions/vision.py [hunk 4]
```diff
@@ -280,3 +281,22 @@ def spectrogram_image(
...
+    clustering_model.fit(
+        np.array(
+            [[ssim(f, ff, multichannel=True) for ff in frames] for f in frames]
```
Comment (changhiskhan) [inline python/rikai/spark/functions/vision.py:299]: assuming ssim is commutative and self-ssim is a constant, you can save 1/2 of the computations here no?

File: python/rikai/spark/functions/vision.py [hunk 5]
```diff
@@ -280,3 +281,22 @@ def spectrogram_image(
...
+        )
+    )
+    return clustering_model.labels_.tolist()
```
Comment (changhiskhan) [inline python/rikai/spark/functions/vision.py:302]: some unit tests?

**Review by changhiskhan — COMMENTED** [review]
looks cool. some comments/questions. thanks for the PR!

File: python/rikai/spark/functions/vision.py (hunk 2 above)
Comment (smellslikeml) [reply python/rikai/spark/functions/vision.py:287]: Just need the dimensions of reference and test images to be the same for SSIM.

File: python/rikai/spark/functions/vision.py (hunk 3 above)
Comment (smellslikeml) [reply python/rikai/spark/functions/vision.py:296]: will update using [skimage.transform.resize](https://scikit-image.org/docs/dev/api/skimage.transform.html#skimage.transform.resize)

File: python/rikai/spark/functions/vision.py (hunk 4 above)
Comment (smellslikeml) [reply python/rikai/spark/functions/vision.py:299]: will do

**Comment by changhiskhan** [comment]
I don't think we can use this as is. I'm going to close this PR and we can create a new one for image clustering as needed
